docs/serialization-design.md

```diff
@@ -0,0 +1,64 @@
+
+All primitives are serialized in little-endian order. By making this choice explicitly, there is no ambiguity across different platforms, and by not allowing alternatives, this makes the serialization suitable for use in Merkle DAGs. A majority of platforms are little-endian, so this is also good for the typical case.
+
+In DyND's storage within memory, there is some metadata which describes memory layout, data references, and other details incidental to how things are stored. For serialization, the layout is defined completely in terms of the type, and there is no such metadata. For example, the `fixed` dimension type has a stride in its metadata. In the serialization of a strided array, the data is always tightly packed, and no stride is stored anywhere.
+
+## Deserialize Callable
+
+The `deserialize` callable has signature `(bytes, dst_tp: type) -> Any`, where the output array has the specified destination type. The generic case of of deserializing a general DyND array is treated as an element of type `array[Any]`, as noted in the serialize section above.
+
+## Type to Bytes Mapping
+
+* `bool`
+  * Single byte `0x00` (false) `0x01` (true)
+* `int##`, `uint##`, `float##`
+  * The bytes are packed in little endian order.
+* `complex[T]`
+  * The real and imaginary components are serialized one after the other.
+* `bytes[N]`
+  * The bytes are stored as is.
+* `bytes`
+  * The size is stored as a [base 128 Varint](https://developers.google.com/protocol-buffers/docs/encoding?hl=en#varints), and the bytes follow as is.
+* `string`
+  * The UTF-8 encoding of the string is serialized identically to `bytes`.
+* `char`
+  * The 32-bit code point is serialized as is. (TODO: Would be better as UTF-8?)
+* `type`
+  * The type represented as a string, serialized as `string`. (TODO: Need to specify a well-defined type to string serialization that we will never change.)
+* `void`
+  * Adds no bytes to the serialization.
+* `N * T`
+  * Each element is serialized in sequence (a packed contiguous serialization). Because `N` is part of the type, it does not need to be stored.
+* `var * T`
+  * The size is stored as a [base 128 Varint](https://developers.google.com/protocol-buffers/docs/encoding?hl=en#varints), and then the elements are serialized the same as `N * T`.
+* `(T0, T1, ...)`, `{field0: T0, field1: T1, ...}`
+  * Each element of the tuple/struct is serialized in sequence (a packed contiguous serialization).
+* `?T`
+  * For types where the missing data sentinel is well-defined, the data is serialized just as `T` would be. For other types, missing is serialized as a single byte `0x00`, and available data is serialized as a single byte `0x01` followed by the serialization as `T`.
+* `pointer[T]`
+  * Serialized exactly as `T` would be.
+* `array[T]`
+  * Serialized as if it were a tuple `(type, T)`, with the type serialized, immediately followed by the data.
```

Comment:
As people add custom types they'll also want a mechanism to serialize and deserialize those types.  What is a good solution here?